### ISSUE TITLE
Support gzipped FASTA files?

### ISSUE THREAD

**Issue by peterjc** [pyani-plus/pyani-plus#214]
## Summary:

We currently expect as input a folder of decompressed FASTA files (see https://github.com/pyani-plus/pyani-plus/blob/v0.0.1a1.dev1/pyani_plus/__init__.py#L42 which accepts ``*.fasta``, ``*.fna`` and ``*.fas``).

It is reasonably common to keep FASTA files gzip compressed, e.g. ``*.fna.gz`` from the NCBI FTP site via https://github.com/kblin/ncbi-genome-download

It might therefore be useful to support gzipped FASTA files directly in pyANI-plus.

## Tool support:

Which of the underlying tools will accept a gzipped FASTA file directly? Sourmash does, but most likely we would need to decompress to ``$TMPDIR`` for the rest - adding to our code complexity.

## Check checksum:

Also, would we want to use the MD5 checksum of ``example.fna.gz`` or ``example.fna``?

e.g. Looking at https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/001/005/405/GCF_001005405.1_ASM100540v1/ the NCBI provide two files,

* https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/001/005/405/GCF_001005405.1_ASM100540v1/md5checksums.txt contains ``458ee9dc28e579c5fe619e510d414be0  ./GCF_001005405.1_ASM100540v1_genomic.fna.gz``

* https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/001/005/405/GCF_001005405.1_ASM100540v1/uncompressed_checksums.txt contains ``./GCF_001005405.1_ASM100540v1_genomic.fna	9987d803716bd9883bd182537835d808	0f94fbc5	10081771`` (file, md5sum, crc32, size).
The ``MD5SUM`` text file downloaded by ``ncbi-genome-download`` has the checksum of ``*.fna.gz`` but mentions ``uncompressed_checksums.txt``

Those match:

```console
$ md5sum GCF_001005405.1_ASM100540v1_genomic.fna.gz 
458ee9dc28e579c5fe619e510d414be0  GCF_001005405.1_ASM100540v1_genomic.fna.gz
$ cat GCF_001005405.1_ASM100540v1_genomic.fna.gz | gunzip | md5sum
9987d803716bd9883bd182537835d808  -
```

I think we should use the later, MD5 of the decompressed file - cross reference #155. However, this would mean decompressing the files as part of the initial scan of a directory when starting any new run (which we could cache if using the local executor, but that might take too much space).

**Comment by kiepczi** [comment]
@peterjc  and I talked about this yesterday, so I thought I’d write down a quick note so we don’t forget what was said.

We both agreed this is going to be a useful feature for `pyani-plus`, especially for handling larger analyses. Unzipping a lot of files takes time and uses up a ton of space, so this could be a much better way to go.

One thing we also mentioned is that some tools, like `nucmer`, don’t accept gzipped files as input. Peter suggested we could unzip the files into a temporary folder, run the analysis, and then delete them afterward.

**Comment by widdowquinn** [comment]
I also think it's a good idea to offer the choice to the user. The trade-offs are probably obvious:

- uncompressed:
  - more storage needed, which may be a problem for large analyses
  - needs no decompression
  - all tools accept the same input
  
- compressed:
  - less storage needed, potentially facilitating larger analyses
  - still needs to be uncompressed _somewhere_, whether within the tool (`sourmash`), by us before analysis (`nucmer`/`BLAST`, maybe), or by the user. That can be done once (by the user) or for each analysis. Once is quicker, overall.
  - not all tools accept compressed files

If offering the ability to work with compressed files, we should warn the user in the documentation about false economies of requiring 1X versus ManyX decompression.

A related issue is network transfer of files: tar/compression of files to move them from the main storage to individual nodes is usually worthwhile (so you only transfer one file, which saves network overhead), decompressing/untarring on the worker node.

In terms of MD5 hashes, will we need both? One to confirm that the `.gz` download did not corrupt the file, and one to confirm that extraction did not corrupt the file?

**Comment by peterjc** [comment]
There are definitely trade offs here.

We would only need the MD5 _after_ decompression. I'm not sure how useful it would be to track the MD5 of the compressed version(s), keeping in mind two gzipped version of the same file need not be identical (e.g. different timestamp in the header, or different compression levels).

**Comment by peterjc** [comment]
Yes. The very first thing we do for a new run is compute the MD5 of all the FASTA files in the input directory (with a progress bar). This can be on-the-fly decompression and MD5 computation, a simple change here:

https://github.com/pyani-plus/pyani-plus/blob/v0.0.1a1.dev2/pyani_plus/utils.py#L65

We *could* at that point decompress the FASTA to disk, but that blows up the disk usage. I think decompression should be restricted to the worker jobs using local temp and deleted after use (especially for remote nodes via SLURM, but perhaps best for local jobs too).

(If we were still using command lines defined within snakemake, we could use piping there to handle the on-the-fly decompression for input to each tool)

**Comment by widdowquinn** [comment]
I agree about handing decompression on to the worker nodes.

I think that we might consider a `tar` step before moving the `.gz` files to worker nodes, too. It will certainly be worth doing for larger comparisons - I took Dundee's cluster out moving too many files, once 😊 

Isn't there a way to pipe using `subprocess.run`? All might not be lost, there.

**Comment by peterjc** [comment]
Good point - decompressing multiple ``example.fasta.gz`` file from networked storage to local ``/tmp/.../example.fasta`` might become a networking issue. Something to keep in mind generally - as even plain FASTA inputs might need staging to each node in bulk?

There is some piping support in ``subprocess`` but I've never used this with two child processes (only writing/reading strings from the Python side).

**Comment by peterjc** [comment]
Closed by #222
